### DISCUSSION TITLE
Custom APT repos as datasource

### DISCUSSION THREAD

**Discussion by felipecrs** [renovatebot/renovate#27716] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

I install Docker CE CLI in my ubuntu-based Dockerfile through the official Docker APT repo:

```dockerfile
FROM ubuntu:22.04

ARG DOCKER_CE_CLI_VERSION=25.0.3

# Install Docker CLI
RUN apt-get update \
  && apt-get install -y apt-transport-https ca-certificates curl gnupg \
  && curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor -o /etc/apt/keyrings/docker.gpg \
  && echo "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu $(. /etc/os-release && echo "${VERSION_CODENAME}") stable" | tee /etc/apt/sources.list.d/docker.list \
  && apt-get update \
  && apt-get install -y --no-install-recommends "docker-ce-cli=5:${DOCKER_CE_CLI_VERSION}-1~ubuntu.22.04~jammy" \
  # Clean up \
  && rm -rf /var/lib/apt/lists/*
```

For now I can pick them from GitHub (https://github.com/moby/moby/releases), but this is not perfect because some releases don't get published into the APT repo, some take several days to be published and so.

All issues and discussions I found ended up with repology, but repology does not keep track of such repositories.

Similarly there's Microsoft's VS Code repository and Google Chrome's.

Anyway, just opening this discussion to hear thoughts. Someone else will probably encounter the same situation someday.

**Comment by felipecrs** [comment]
Another idea I have to work around the limitation of fetching the versions from GitHub is to add a hook to check if the final file actually exists on the APT repo, if it does not, do not propose the update:

```bash
github_version="25.0.3"
apt_version="${github_version}-1~ubuntu.22.04~jammy"

curl -fsSL -o /dev/null -w "%{http_code}" "https://download.docker.com/linux/ubuntu/dists/jammy/pool/stable/amd64/docker-ce-cli_${apt_version}_amd64.deb"
```

But `postUpgradeTasks` can't be used with the Mend App. Perhaps some new built-in `postUpgradeOption` like `checkUrl` where we can pass something like `checkUrlTemplate: "https://download.docker.com/linux/ubuntu/dists/jammy/pool/stable/amd64/docker-ce-cli_{{{ newVersion }}}-1~ubuntu.22.04~jammy_amd64.deb"`

**Reply by felipecrs** [reply]
It should also be possible to just get a list of files from https://download.docker.com/linux/ubuntu/dists/jammy/pool/stable/amd64/, filter out those which match the depName, apply some versioning customization and sort them.

**Reply by felipecrs** [reply]
Or better yet, just setup the repo and run apt commands to figure the list of available versions like `apt-cache madison docker-ce-cli`.

**Comment by cospeedster** [comment, marked as answer]
Maybe you can try the new [deb datasource](https://docs.renovatebot.com/modules/datasource/deb/) which was introduced in release [38.57.0](https://github.com/renovatebot/renovate/releases/tag/38.57.0).

**Reply by zisuu** [reply]
Does someone have a working example configuration to share for this use case?

**Reply by zisuu** [reply]
I was able to get a PR with this packageRule:
```json
{
  "matchDatasources": ["deb"],
  "matchPackageNames": ["docker-ce-cli"],
  "registryUrls": [
    "https://download.docker.com/linux/ubuntu?suite=stable&components=stable&release=jammy&binaryArch=amd64"
  ]
}
```
but my issue is that the semantic version tag format I want to use (27.3.1) get's screwed by Renovate:
![image](https://github.com/user-attachments/assets/7ecee6f6-455d-450c-ac33-8c38b2e88eda)


I also tried with `"versioning": "loose"` and `"extractVersion": "^\\d+:?(\\d+\\.\\d+\\.\\d+)"` but that did not work either.

**Reply by rarkins** [reply]
Try versioning=semver

**Reply by zisuu** [reply]
thanks @rarkins but I had no luck:
```bash
INFO: Found no results from datasource that look like a version (repository=xx/devplat-java-devcontainer-playground, dependency=docker-ce-cli)
   "result": {
     "releases": [],
     "homepage": "https://www.docker.com",
     "registryUrl": "https://download.docker.com/linux/ubuntu?suite=stable&components=stable&release=jammy&binaryArch=amd64"
   }
```

**Reply by rarkins** [reply]
So what are the string patterns returned from the datasource? Are all acceptable only certain patterns? You'll need to extract them

**Reply by rarkins** [reply]
Actually, please start a new discussion about this, do not continue using this one
